### PR TITLE
Adding performance counter type information

### LINKED ISSUE #4438: Add performance counter flag is monotically increasing
Visualization tools might want to present monotically increasing counters differently. We should add some way to distinguish those from instantaneous values.

### PR DESCRIPTION
**Pull request by hkaiser**
Fixes #4438

This PR assigns appropriate types (aggregating, monotonicaly increasing, average count, or average timer, etc.) to all performance counters.

One of API function overloads `install_counter_type` has gained an additional (optional) argument. Recompilation will work, however. Thus this PR introduces an ABI break.

### CODE REVIEW

**Comment by hkaiser** [comment]
@khuck For now, I have added a new overload for `external_timer::sample_value` that is called with a `counter_info` instance (https://github.com/STEllAR-GROUP/hpx/pull/4441/files#diff-0115424e11d4f55fdb03f6b82d4e22f8R31-R38). This currently just forwards to the old overload by passing along the counter name (as before). Once you have added support to APEX to introspect the counter type (it's a member of the `counter_info`) you will have to adapt this new overload accordingly.

**Comment by hkaiser** [comment]
@msimberg I think this one is good to be merged now.

**Comment by msimberg** [comment]
> @khuck For now, I have added a new overload for `external_timer::sample_value` that is called with a `counter_info` instance (https://github.com/STEllAR-GROUP/hpx/pull/4441/files#diff-8a3dc9aad233c09844e8b00e3be76567R32-R39). This currently just forwards to the old overload by passing along the counter name (as before). Once you have added support to APEX to introspect the counter type (it's a member of the `counter_info`) you will have to adapt this new overload accordingly.

Just to check, this is not meant to be compatible with the current APEX, right? (https://cdash.cscs.ch//viewBuildError.php?buildid=100269)

File: libs/performance_counters/include/hpx/performance_counters/apex_sample_value.hpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#include <string>
+
+namespace hpx { namespace util {
+
+    // The actual function pointers. Some of them need to be exported,
+    // because through the miracle of chained headers they get referenced
+    // outside of the HPX library.
+    static inline void sample_value(const std::string& name, double value)
+    {
+        if (sample_value_function != nullptr)
```
Comment (msimberg) [inline libs/performance_counters/include/hpx/performance_counters/apex_sample_value.hpp]: Maybe?
```suggestion
        if (external_timer::sample_value_function != nullptr)
```

Comment (hkaiser) [reply libs/performance_counters/include/hpx/performance_counters/apex_sample_value.hpp]: Uhh, the whole code block belongs into the `namespace external_timer`, thanks for spotting this! Will fix asap.

**Comment by hkaiser** [comment]
> Just to check, this is not meant to be compatible with the current APEX, right? (https://cdash.cscs.ch//viewBuildError.php?buildid=100269)

The intention was for it to be compatible. The problem is that it misses the `namespace external_timer` surrounding the new code. Will fix asap.

File: libs/performance_counters/include/hpx/performance_counters/apex_sample_value.hpp (hunk 1 above)
Comment (hkaiser) [reply libs/performance_counters/include/hpx/performance_counters/apex_sample_value.hpp]: That should be fine now.

**Comment by khuck** [comment]
OK, I had to make some changes to APEX.  When these changes are merged, APEX needs to merge branch https://github.com/khuck/xpress-apex/tree/fixes_for_hpx_4441.

Here's some sample test output to prove that it worked...(or that the previous functionality isn't broken)
```
[khuck@delphi hpx-Debug]$ /home/users/khuck/src/phylanx/tools/buildbot/build-delphi-x86_64-Linux-gcc/hpx-Debug/bin/apex_action_count_test --hpx:threads=1 --hpx:print-counter=/threads{locality#0/worker-thread#*}/count/cumulative
fibonacci(10) == 55
elapsed time: 0.033328 [s]
/threads{locality#0/worker-thread#0/total}/count/cumulative,1,0.070769,[s],188
/threads{locality#0/pool#default/worker-thread#0}/count/cumulative,1,0.070846,[s],189

Elapsed time: 0.0794527 seconds
Cores detected: 72
Worker Threads observed: 1
Available CPU time: 0.0794527 seconds

Counter                                   : #samples | minimum |    mean  |  maximum |  stddev 
------------------------------------------------------------------------------------------------
/threads{locality#0/pool#default/worke... :        1    189.000    189.000    189.000      0.000   
/threads{locality#0/worker-thread#0/to... :        1    188.000    188.000    188.000      0.000   
                    1 Minute Load average :        1      2.540      2.540      2.540      0.000   
------------------------------------------------------------------------------------------------

Timer                                                : #calls  |    mean  |   total  |  % total  
------------------------------------------------------------------------------------------------
                                           <unknown> :        6      0.000      0.000      0.541
                                           APEX MAIN : Not called since reset.
                        async_launch_policy_dispatch :        5      0.001      0.004      5.530
            broadcast_call_shutdown_functions_action :        2      0.000      0.000      0.593
                      call_shutdown_functions_action :        2      0.003      0.006      7.052
                   create_performance_counter_action :        2      0.001      0.001      1.731
                                    fibonacci_action :      174      0.000      0.029     36.622
   hpx::parallel::execution::parallel_executor::post :        2      0.000      0.000      0.093
                              load_components_action :        1      0.012      0.012     15.224
                   primary_namespace_colocate_action :        2      0.000      0.000      0.427
                                          run_helper :        1      0.017      0.017     21.520
                                 shutdown_all_action :        1      0.002      0.002      1.992
                                           APEX Idle :                          0.007      8.676
------------------------------------------------------------------------------------------------
                                        Total timers : 198
Calls to fibonacci_action: 177
APEX measured calls to fibonacci_action: 174
[khuck@delphi hpx-Debug]$ 
```

**Comment by khuck** [comment]
@hkaiser here's a summary of the changes: https://github.com/khuck/xpress-apex/commit/39f0cb79a42ba987234f26b3a5c6139e1bb8a45d

**Comment by hkaiser** [comment]
> @hkaiser here's a summary of the changes: khuck/xpress-apex@39f0cb7

Those APEX changes should be compatible and should not break with older versions of HPX. This version of HPX will not build against older versions of APEX, however. I think we're fine merging this PR. Thanks!

**Comment by msimberg** [comment]
@hkaiser I'll let you merge this one. I'm not sure how to interpret

> The intention was for it to be compatible.

versus

> This version of HPX will not build against older versions of APEX

Pycicle still says [this](https://cdash.cscs.ch//viewBuildError.php?buildid=100423) but I'm not sure what you mean by older versions of APEX.

**Comment by hkaiser** [comment]
> @hkaiser I'll let you merge this one. I'm not sure how to interpret
> 
>> The intention was for it to be compatible.
> 
> versus
> 
>> This version of HPX will not build against older versions of APEX
> 
> Pycicle still says this but I'm not sure what you mean by older versions of APEX.

I was thinking while writing and in the end I contradicted myself ;-)

This PR removes the performance counter stubs (as we wanted to get rid of stubs over time). I didn't think anybody would depend on them, honestly, but APEX apparently did. 

Pycicle has to fail as APEX needs to be updated to not use the stubs anymore. @khuck please go ahead and merge your patch first. We can then verify that this PR is fine. We might have to update it to refer to another APEX version before we retrigger the build, however.

**Comment by msimberg** [comment]
> > @hkaiser I'll let you merge this one. I'm not sure how to interpret
> > > The intention was for it to be compatible.
> > 
> > 
> > versus
> > > This version of HPX will not build against older versions of APEX
> > 
> > 
> > Pycicle still says this but I'm not sure what you mean by older versions of APEX.
> 
> I was thinking while writing and in the end I contradicted myself ;-)
> 
> This PR removes the performance counter stubs (as we wanted to get rid of stubs over time). I didn't think anybody would depend on them, honestly, but APEX apparently did.
> 
> Pycicle has to fail as APEX needs to be updated to not use the stubs anymore. @khuck please go ahead and merge your patch first. We can then verify that this PR is fine. We might have to update it to refer to another APEX version before we retrigger the build, however.

Thanks, now it makes sense :) I'm also fine with merging this first. If we wait for APEX to be updated it should get a tag, and this PR should update the tag we use internally (I can force it on pycicle, but it needs to be changed here anyway).

**Comment by khuck** [comment]
Sorry for the delay!  Yes, the changes in APEX are backwards compatible with current HPX.  You are good to go!  I've done a release and tagged APEX regardless.  The latest APEX release is now v2.1.8.  Thanks!

**Comment by msimberg** [comment]
@hkaiser note that I pushed a commit to update the APEX tag.

**Comment by hkaiser** [comment]
> @hkaiser note that I pushed a commit to update the APEX tag.

Did the clang-apex builder pick up the tag? Or has @khuck missed places to fix?

**Comment by msimberg** [comment]
> > @hkaiser note that I pushed a commit to update the APEX tag.
> 
> Did the clang-apex builder pick up the tag? Or has @khuck missed places to fix?

I think it did pick up the new tag. The number of build errors reduced significantly. Looks like the errors come from our own APEX tests: https://cdash.cscs.ch//viewBuildError.php?buildid=100577.

**Comment by hkaiser** [comment]
>>> @hkaiser note that I pushed a commit to update the APEX tag.
>> 
>> Did the clang-apex builder pick up the tag? Or has @khuck missed places to fix?
> 
> I think it did pick up the new tag. The number of build errors reduced significantly. Looks like the errors come from our own APEX tests: https://cdash.cscs.ch//viewBuildError.php?buildid=100577.

Right, I missed that part. I'll look into fixing the test.

File: examples/apex/apex_policy_engine_active_thread_count.cpp [hunk 2]
```diff
@@ -57,35 +55,29 @@ std::uint64_t fibonacci(std::uint64_t n)
...
 }
 
 void setup_counters() {
     try {
-        id_type id = get_counter_id();
+        performance_counter counter = get_counter();
         // We need to explicitly start all counters before we can use them. For
         // certain counters this could be a no-op, in which case start will
         // return 'false'.
         performance_counter::start(hpx::launch::sync, id);
```
Comment (msimberg) [inline examples/apex/apex_policy_engine_active_thread_count.cpp]:
```suggestion
        counter.start(hpx::launch::sync, id);
```
I think?

Comment (hkaiser) [reply examples/apex/apex_policy_engine_active_thread_count.cpp]: Right... I missed that spot (I can't compile it here on my machine), thanks.

**Pull request merged**